### ISSUE TITLE
Enable org-protocol support out of the box

### ISSUE THREAD

**Issue by jkp** [d12frosted/homebrew-emacs-plus#285]
It would be great to have the org-protocol URL handler registered out of the box to enable org-capture to work.

This is supported by another mac-friendly port:  [emacs-mac](https://github.com/railwaycat/homebrew-emacsmacport) or by installing a [dedicated utility](https://github.com/sprig/org-capture-extension) however the last method seems to be broken on newer versions of the OS.

I saw that this was mentioned in #158 but I thought it would be worth adding a dedicated feature request here to track progress independently of fixing that issue.

**Comment by kickingvegas** [comment]
Any progress to this issue?

**Comment by d12frosted** [comment]
@kickingvegas I have not worked on this feature. PRs are welcome :)

**Comment by kickingvegas** [comment]
@d12frosted - How about this counter-proposal - I&rsquo;d be happy to pair code this PR with you! I&rsquo;m asking because I&rsquo;m really not well setup to build Emacs plus for Homebrew. Looking at the two 28.2 repos, it seems like the chief differences are in two files:

-   `org-protocol.el`
-   `Info.plist.in`

Repos
- <https://bitbucket.org/mituharu/emacs-mac/src/master/>
- <https://ftp.gnu.org/gnu/emacs/emacs-28.2.tar.xz>

diff of [emacs-mac/lisp/org/org-protocol.el](https://bitbucket.org/mituharu/emacs-mac/src/master/lisp/org/org-protocol.el) with [emacs-28.2/lisp/org/org-protocol.el](https://github.com/emacs-mirror/emacs/blob/emacs-28/lisp/org/org-protocol.el)

```text
631c631
< (defun org-protocol-check-filename-for-protocol (fname restoffiles client)
---
> (defun org-protocol-check-filename-for-protocol (fname restoffiles _client)
672,675c672
< 		  ;; Emacs Mac port directly handles `org-protocol'
< 		  ;; URLs without the help of external commands or
< 		  ;; apps.  In this case, `client' is set to nil.
<                   (when (and client (plist-get (cdr prolist) :kill-client))
---
>                   (when (plist-get (cdr prolist) :kill-client)

```

The emacs-mac port makes extensive changes to the `Info.plist.in` file which is used to generate the `Info.plist` that is put in app bundle. That said I think the most significant difference with respect to Org Protocol support is adding `org-protocol` to the `CFBundleURLTypes` key as shown below.

- [emacs-mac/mac/templates/Info.plist.in](https://bitbucket.org/mituharu/emacs-mac/src/master/mac/templates/Info.plist.in) 

- [emacs-28/nextstep/templates/Info.plist.in](https://github.com/emacs-mirror/emacs/blob/emacs-28/nextstep/templates/Info.plist.in)


```xml
<key>CFBundleURLTypes</key>
<array>
  <dict>
    <key>CFBundleURLName</key>
    <string>Email Address URL</string>
    <key>CFBundleURLSchemes</key>
    <array>
      <string>mailto</string>
    </array>
  </dict>
  <dict>
    <key>CFBundleURLName</key>
    <string>Org Protocol URL</string>
    <key>CFBundleURLSchemes</key>
    <array>
      <string>org-protocol</string>
    </array>
  </dict>
</array>
```

**Comment by yaqubroli** [comment]
Any progress on this? If this hasn't just been incorporated silently into master and not referenced in the issue, then I can make the PR

**Comment by d12frosted** [comment]
My bad. It slipped out of my head. @kickingvegas I will review your comment after I return home and get access to my laptop.

**Comment by maikol-solis** [comment]
Is there any update on this? Thanks.

**Comment by d12frosted** [comment]
@maikol-solis there are no updates; I would really apreciate someone just figuring out how to solve this and sending a PR.

**Comment by kickingvegas** [comment]
Found the commit that introduced org-protocol support in the [Mitsuharu fork](https://bitbucket.org/mituharu/emacs-mac/src/master/): 0d4110bb05f080a3f4f07ae9338d4d61b4d29c16

Affected files afaik:

    lisp/org/org-protocol.el
    lisp/term/mac-win.el
    mac/templates/Info.plist.in

The function `mac-ae-get-url` is what handles a passed URL to the Emacs Mac App from an external app.

    commit 0d4110bb05f080a3f4f07ae9338d4d61b4d29c16
    Author: YAMAMOTO Mitsuharu <mituharu@math.s.chiba-u.ac.jp>
    Date:   Thu Jan 29 12:59:45 2015 +0900
    
        * emacs-24.4-mac-5.3 (2015-01-29)
    
        Blend-and-blur of background color on OS X 10.10 is temporarily
        disabled to avoid mode-line erasure on frame focus switch.  For some
        reason, mode-line row's `enabled_p' flag in the current matrix is
        sometimes reset even if it is visible.  This causes blank mode-line on
        "expose" (drawRect:) events.  In principle, this mode-line erasure can
        also happen on older versions of OS X or this version of Emacs Mac
        port.  But it is usually observed on OS X 10.10 together with previous
        5.x versions of Emacs Mac port, because the use of NSVisualEffectView,
        which provides the blend-and-blur feature, triggers full frame
        "exposure" on focus switch.
    
        ** Fixed bugs
    
        *** Special-display completion frame becomes transparent.
        Reported by Alan Shumitt.
        Adapt a fix for Bug#16619.
    
        *** `mac-auto-ascii-mode' on TTY terminals disables prefix keys with
        the meta key.
    
        *** Control-F2 is passed to the system even if
        mac-pass-control-to-system is set to nil.
    
        *** Synthetic bold text looks thinner on Retina display/HiDPI mode.
        This is because stroke line width for text drawing is not correctly
        scaled (whereas it is correctly scaled when drawn to bitmaps) on such
        environments.  I've been reporting this to Apple on every OS X update
        since it was 10.7.4, but I couldn't get any response until recently.
        At last Apple answers that there are no plans to address this issue
        (rdar://11644870) currently.  So I added a workaround that would cause
        a bad reverse effect if the original issue were fixed.
    
        *** `mac-ae-set-reply-parameter', which is used for reporting an error
        in Apple event hander, has not been working since emacs-24.2-mac-3.2.
    
        *** Input source functions ignore disabled input source ID strings.
    
        ** Improvements
    
        *** URL handler via Apple events now accepts the `org-protocol' scheme
        as well as `mailto'.

-   Notes <span class="timestamp-wrapper"><span class="timestamp">[2023-12-02 Sat 15:10]</span></span>

Yantar92 posted this on IRC about Org-Protocol support in Core. Apparently it is very difficult to abstract this behavior away to support multi-platform.

**Comment by pastnontra** [comment]
Emacs plus has provided Emacs Client.app, and it would be nice to see org-protocol support out of the box, too. To be honest, org-protocol is the biggest reason why I gonna to use Emacs Client.app when Emacs.app handles frame reuse pretty well for my use case.

**Comment by d12frosted** [comment]
@pastnontra agreed, will try bump priority of this one;

meanwhile I would appreciate feedback on Emacs Client.app - does it work for you?

**Comment by pastnontra** [comment]
Yes, I think so. But the frame reuse strategy is a bit confusing. I'll do further test and feedback in #158.

I'd also like to mention I try to do org-protocol by editing main script myself, I follow the [Org roam manual](https://www.orgroam.com/manual.html#org_002droam_002dprotocol), which seems the most completed instructions. But somehow it can't work, the Emacs Client flash and gone without any feedback.

**Comment by kickingvegas** [comment]
FWIW, since the time I last posted here, I wrote and published two macOS apps, Scrim and Captee, to support Org protocol as a workaround to native Emacs support for it. The first app (Scrim) acts as an Org protocol proxy, processing any 3rd party program that makes a URL scheme request to `org-protocol://`. The second app (Captee) will construct an Org protocol request using the macOS Share menu.

Here's a link to them on the macOS App Store.
<https://itunes.apple.com/us/app-bundle/id1810494622?mt=12>

**Comment by eamonnsullivan** [comment]
I came here to subscribe, but just want to mention the [Org Roam instructions](https://www.orgroam.com/manual.html#Mac-OS-1) again. They *did* work for me, so possibly something has changed recently. I'm using emacs-plus@31 and was looking for a fix, since Emacs 31 [breaks Scrim](https://github.com/kickingvegas/scrim/discussions/74).

**Comment by d12frosted** [comment]
closed too soon, need to double test everything

**Comment by d12frosted** [comment]
org-protocol support is now available in Emacs Client.app 🎉

Starting with the next build of `emacs-plus@30` and `emacs-plus@31`, `Emacs Client.app` handles `org-protocol://` URLs out of the box.

#### Setup

1. Add to your Emacs init file:
   ```lisp
   (require 'org-protocol)
   ```
2. Copy `Emacs Client.app` to `/Applications` for reliable URL handling:
   ```
   cp -r "$(brew --prefix)/opt/emacs-plus@30/Emacs Client.app" /Applications/
   ```
3. Install a browser extension like https://github.com/sprig/org-capture-extension or configure your browser bookmarklets.

#### How it works

Emacs Client.app now registers the org-protocol URL scheme with macOS and includes an AppleScript handler that passes URLs to emacsclient. When you click an org-protocol:// link (from a browser extension, bookmarklet, or any app), macOS routes it to Emacs Client.app, which forwards it to your running Emacs daemon.

### Testing

You can test with:
`open "org-protocol://capture?template=t&url=https://example.com&title=Test"`

For org-roam users, see the https://www.orgroam.com/manual.html#org_002droam_002dprotocol for protocol setup.

---
This has been a long-requested feature (4 years! 😅). Thanks to everyone who contributed research and suggestions in this thread, especially @kickingvegas for the detailed analysis of the `emacs-mac` implementation.
